### ISSUE TITLE
OWS is not able to handle multi release jars

### ISSUE THREAD

**Issue by AG72** [karakun/OpenWebStart#384]
I'm struggeling with a legacy Java 8 application that shall now run with Java 11 and OWS. Starting the application from within my IDE, I found several `--add-exports` and `--add-opens` that I need to add so that the application will start properly with Java 11. 

I then added the same add-exports/add-opens to my JNLP file this way:
```
<j2se version="11+" java-vm-args="--add-exports=java.desktop/sun.swing=ALL-UNNAMED --add-exports=java.desktop/sun.swing.table=ALL-UNNAMED   --add-exports=java.desktop/sun.swing.plaf.synth=ALL-UNNAMED   --add-opens=java.desktop/javax.swing.plaf.synth=ALL-UNNAMED   --add-opens=java.desktop/javax.swing.plaf.basic=ALL-UNNAMED   --add-opens=java.desktop/javax.swing=ALL-UNNAMED   --add-opens=java.desktop/javax.swing.tree=ALL-UNNAMED   --add-opens=java.desktop/java.awt.event=ALL-UNNAMED   --add-exports=java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED   --add-exports=java.desktop/sun.awt.shell=ALL-UNNAMED   --add-exports=java.base/java.lang=ALL-UNNAMED   --add-opens=java.base/java.lang=ALL-UNNAMED   --add-exports=java.desktop/com.sun.awt=ALL-UNNAMED   --add-exports=java.base/sun.security.action=ALL-UNNAMED   --add-opens=java.base/sun.security.action=ALL-UNNAMED   --add-opens=java.base/java.security=ALL-UNNAMED   --add-opens=java.base/java.util.concurrent=ALL-UNNAMED   --add-opens=java.desktop/java.beans=ALL-UNNAMED"/>      
```

But when started from OWS it crashes with an exception, which if starting from within my IDE is gone as soon as the mentioned add-exports/add-opens are in place!
```
	...
	Caused by: java.lang.NoClassDefFoundError: sun/swing/plaf/synth/SynthIcon
	...
```
	
At first, I thought those would not be handed over to the application in OWS. But I found a closed ticket (OpenWebStart rejects/discards "--add-opens" java argument (Java 11) #322) that this was fixed with 1.3.0 release. And in the stage1 log file, at the end I can see that all parameters were set as I want them to be.
 I then started the application by hand with the command line its  parameters extracted from the end of stag1 log:

```
"C:\Program Files\Java\adopt_jdk-11.0.9+11\bin\java.exe" "-Xbootclasspath/a:C:\Program Files\OpenWebStart\openwebstart.jar" --add-exports=java.desktop/sun.swing=ALL-UNNAMED --add-exports=java.desktop/sun.swing.table=ALL-UNNAMED ... ...  net.sourceforge.jnlp.runtime.Boot -Xnofork c:\tmp\neo6.jnlp
```

and crashed with the above exception again. In the resulting ITW log file, I can not see any hint for the problem, so I wonder why the add-exports/add-opens parameters seem not to be considered? In the ITW log, I can see that the JNLP file is parsed again and a new class loader is established directly afterwards, maybe this is related to the problem?

**Comment by sclassen** [comment]
So far we have not had other user report problems with the java module system.
Can you attache the JNLP file to this issue such that we can try to reproduce the issue.

**Comment by AG72** [comment]
By the way, the `illegal-access` command line option is not recognized by OWS. But it is listed in https://docs.oracle.com/javase/9/tools/java.htm#JSWOR624 - so I think it should by whitelisted by default?
When I use it I get:
[ITW-CORE][2021-01-26 11:35:54.562 MEZ][WARN ][net.sourceforge.jnlp.Parser] Ignoring java-vm-args due to illegal Property --illegal-access=permit
and it is not handed over from stage 1 to stage 2.

**Comment by AG72** [comment]
And here is the JNLP file:
[OWS_neo6.zip](https://github.com/karakun/OpenWebStart/files/5872897/OWS_neo6.zip)

Thanks in advance!

**Comment by sclassen** [comment]
Because of security considerations we only whitelist a minimal set of vm-args.
You can add additional args to the whitelist in the deployment.properties.

See https://openwebstart.com/docs/OWSGuide.html#_whitelist_for_jvm_arguments

**Comment by sclassen** [comment]
the jnlp looks ok. nothing special.
unfortunately we cannot launch the app to reproduce the behavior

**Comment by AG72** [comment]
> unfortunately we cannot launch the app to reproduce the behavior

I am not allowed to attach our app in here, but I got an OK to send our .war file to you via email. If this is fine for you, please tell me the address to send it to.

**Comment by sclassen** [comment]
openwebstart@karakun.com

Mention the issue number in the subject. Thanks

**Comment by AG72** [comment]
Meanwhile I followed another hint and tried adding the add-exports/add-opens to the main jar MANIFEST.MF - this does not work either.
```
Add-Opens: java.desktop/javax.swing.plaf.synth java.desktop/javax.swin
 g.plaf.basic java.desktop/javax.swing java.desktop/javax.swing.tree j
 ava.desktop/java.awt.event java.base/java.lang java.base/sun.security
 .action java.base/java.security java.base/java.util.concurrent java.d
 esktop/java.beans
Add-Exports: java.desktop/sun.swing java.desktop/sun.swing.table java.
 desktop/sun.swing.plaf.synth java.desktop/com.sun.java.swing.plaf.win
 dows java.desktop/sun.awt.shell java.desktop/com.sun.awt java.base/su
 n.security.action java.base/java.lang
```

**Comment by sclassen** [comment]
OK we finally found some time and are able to reproduce your issue.
We are on it but it looks like this may not be trivial

**Comment by sclassen** [comment]
looks like this could be the issue:
https://twitter.com/lucianct2k2/status/1263496281806209032

> sun.swing.plaf.synth.SynthIcon has been moved to javax.swing.plaf.synth.SynthIcon (probably between 8 and 9)

**Comment by sclassen** [comment]
and this rabbit hole gets deeper...
Because the synthetica.jar is a multi-release jar which has a special class to handle the move of the SynthIcon.
So the problem is that OpenWebStart is currently not capable of handling multi-release jars

**Comment by janakmulani** [comment]
The issue of OWS and multi release jar is being tracked internally. 

You are using Synthetica 2.31.1 which is a multi release jar that works with both Java 8 and Java 9+. Is there a possibility of using Synthetica jar which is not multi release and works exclusively with Java 9+, for example version 3.x jar?

**Comment by AG72** [comment]
We have to support clients using Java 8 and Java 11 in parallel. Right now, all clients are running Java 8. We want to migrate them step by step to Java 11, and then finally cut Java 8 support when all got migrated. So, we either have to use the multi release jar, or have to have separate clients for download for Java 8 / 11 (I'll check internally if this might be an option).

**Comment by sclassen** [comment]
you can also specify jars in the JNLP to only be loaded in case of 11 or 1.8

**Comment by AG72** [comment]
> you can also specify jars in the JNLP to only be loaded in case of 11 or 1.8

Thanks for this hint. Is there an additional feature in OWS, or do you mean by writing something like:
```
	<resources arch="amd64">
		<java version="11*" vendor="AdoptOpenJDK" />
                <jar href="whatever.jar" version="1"/>
	</resources>
```

**Comment by sclassen** [comment]
no additional feature, OWS is following the JNLP standard here.

From the JNLP specification:
```
<java version="1.3" initial-heap-size="64m"/>
<java version="1.2">
  <resources> ... </resources>
</java>
```

**Comment by AG72** [comment]
Following your hint that synthetica.jar being a multi-release jar is the problem, I recomposed this jar to a second "j9" version moving the java9 class from the multiversion branch replacing the original java 8 class in its  standard location. Using this "j9" jar and OWS with Java 11 is now working smoothly :-) . 
So I am on my way creating a running solution by having two downloads, one for Java 8 with original jar, and one for Java 11 with "j9" jar.


Using Stephans idea from his last comment did not work for me. I changed the launch.jnlp to contain:
```
<resources>
      <java version="11">
		  <resources>
			<jar href="synthetica9.jar" version="2.0.0.1-SNAPSHOT"/>
		  </resources>
	  </java>
      <java version="1.8">
		  <resources>
			<jar href="synthetica.jar" version="2.0.0.1-SNAPSHOT"/>
		  </resources>
	  </java>
  </resources>
```

and the version.xml to contain:

```
<resource>
    <pattern>
      <name>synthetica-j9.jar</name>
      <version-id>2.0.0.1-SNAPSHOT</version-id>
    </pattern>
    <file>synthetica-2.31.1-j9.jar</file>
  </resource>
  <resource>
    <pattern>
      <name>synthetica.jar</name>
      <version-id>2.0.0.1-SNAPSHOT</version-id>
    </pattern>
    <file>synthetica-2.31.1.jar</file>
  </resource>
```

It works fine with Java 11, but Java 8 struggeles and also uses "j9" jar so it can not start up saying
```
Caused by: java.lang.ClassNotFoundException: de.javasoft.plaf.synthetica.SyntheticaIcon
	at net.sourceforge.jnlp.runtime.classloader.JNLPClassLoader.lambda$loadClass$9(JNLPClassLoader.java:1445)
```
Interestingly, when I try using Oracle Java 8 and Oracle Webstart (which our customers have right now until we migrated them to OWS) it works.
Also, when I just remove the new `<java version="11">...</java>` section from jnlp file, it works with OWS and Java 8.

**Comment by sclassen** [comment]
looks like you found another bug. Can you open a new issue about OWS selecting the wrong resource if in this usecase. thanks

**Comment by sclassen** [comment]
fixed in https://github.com/AdoptOpenJDK/IcedTea-Web/pull/797
